File: Floral/vector.cpp

#include "vector.h"
#include <stdio.h>
#include "constants.h"

float Vec::PI = 3.1415926f;

Vec::Vec(){
    this->x=1.0f;
    this->y=0.0f;
}
Vec::Vec(GLfloat x, GLfloat y){
    this->x=x;
    this->y=y;
}
Vec Vec::add(Vec v2){
    Vec v(this->x+v2.x, this->y+v2.y);
    return v;
}
Vec Vec::sub(Vec v2) {
	Vec v(this->x - v2.x, this->y - v2.y);
	return v;
}
Vec Vec::neg(){
    Vec v(-(this->x), -(this->y));
    return v;
}
Vec Vec::norm(){
    Vec v;
    GLfloat m=this->mag();
    if (m==0){
        fprintf(stderr, "cal norm for zero length vector");
        exit(1);
    }
    return this->mul(1/m);
}
Vec Vec::mul(GLfloat s){
    Vec v(this->x*s,this->y*s);
    return v;
}

GLfloat Vec::mag(){
    return sqrt(this->x*this->x+this->y*this->y);
}

GLfloat Vec::angle(){
    if (x==0 && y>=0) return 90.0f;
    if (x==0 && y<0) return 270.0f;
    if (x>0 && y==0) return 0.0f;
    if (x<0 && y==0) return 180.0f;
    if (y>0 && x>0){//0-90 deg
		//printf("y=%f, x=%f, y>0,x>0 y/x is %f, atan(y/x) is %f, rad2deg(atan) is %f\n", y, x, y / x, atan(y / x), rad2deg(y / x));
        return Vec::rad2deg(atan(y/x));
    }
    if (y>0 && x<0){//90-180 deg y/x<0 atan<0
        return Vec::regulizeDeg(rad2deg(atan(y/x))+180.0f);
    }
    if (y<0 && x<0){//180-270 deg y/x>0 atan>0
        return Vec::regulizeDeg(rad2deg(atan(y/x))+180.0f);
    }
    if (y<0 && x>0){//270-360 deg y/x<0 atan<0
        return Vec::regulizeDeg(rad2deg(atan(y/x))+360.0f);
    }
    return rad2deg(atan(y/x));
}

File: Floral/lua/Vector.lua
function regulizeDeg(x)
  while (x>360)
  do
    x=x-360
  end
  while (x<0)
  do
    x=x+360
  end
  return x;
end

function rad2deg(x)
  return regulizeDeg(x/math.pi*180.0)
end
function deg2rad(x)
  return regulizeDeg(x)*math.pi/180.0
end

Vector={}
do
   local meta = {
        _metatable = "Private metatable",
        _DESCRIPTION = "Vectors in 2D"
    }

    meta.__index = meta

    function meta:__add( v )
        return Vector(self.x + v.x, self.y+ v.y)
    end
    
    function meta:__sub( v )
        return Vector(self.x - v.x, self.y - v.y)
    end
    function meta:__mul( v )
        return self.x * v.x + self.y * v.y
    end
    
    function meta:__tostring()
        return ("<%g, %g>"):format(self.x, self.y)
    end
    --rotate counter clockwise
    function meta:rotate(deg)
        local rad=deg2rad(deg)
        local sin=math.sin(rad)
        local cos=math.cos(rad)
        return Vector(self.x*cos-self.y*sin,self.y*cos+self.x*sin)
    end
      
    function meta:mag()
        return math.sqrt( self * self )
    end
    
  
    function meta:angle()
        if (self.x==0 and self.y>=0)then return 90.0 end
        if (self.x==0 and self.y<0) then return 270.0 end
        if (self.x>0 and self.y==0) then return 0.0 end
        if (self.x>0 and self.y==0) then return 180.0 end
        
        return rad2deg(math.atan2(self.y,self.x))
    end
    
    function meta:plain()
      return self.x, self.y
    end
    setmetatable( Vector, {
        __call = function( V, x ,y ) return setmetatable( {x = x, y = y}, meta ) end
    } )
end

Vector.__index=Vector

v=Vector(0,1)
print(v:rotate(90))
return Vector
